src/repr/src/row.rs

```diff
@@ -1193,6 +1350,172 @@ impl RowPacker<'_> {
+    ///   that it is possible to generate semantically empty ranges with this
+    ///   function, but they will _not_ be output as empty ranges.
+    /// - To construct ranges without providing the `Datum`s directly, see
+    ///   [`Self::push_range_with`].
+    ///
+    /// # Panics
+    /// - If `lower` and `upper` are both non-null and are of different types.
+    pub fn push_range<'a, D>(
+        &mut self,
+        lower: D,
+        lower_inc: bool,
+        upper: D,
+        upper_inc: bool,
+    ) -> Result<(), RangeError>
+    where
+        D: Borrow<Datum<'a>>,
+    {
+        let lower = lower.borrow();
+        let upper = upper.borrow();
+
+        self.push_range_with(
+            RangeBoundDesc {
+                inf: matches!(lower, Datum::Null),
+                inc: lower_inc,
+            },
+            RangeBoundDesc {
+                inf: matches!(upper, Datum::Null),
+                inc: upper_inc,
+            },
+            |row| {
+                for bound in [lower, upper] {
+                    if !matches!(bound, Datum::Null) {
+                        row.push(bound);
+                    }
+                }
+            },
+        )
+    }
+
+    /// Pushes a `DatumRange` built from the specified arguments, and
+    /// permitting the user to push values into the row using `f`.
+    ///
+    /// - If an error occurs, the packer's state will be unchanged.
+    /// - Prefer using [`Self::push_range`], which prevents most potential
+    ///   panics.
+    /// - Infinite boundaries should be expressed using [`RangeBoundDesc`]'s
+    ///   `inf` field.
+    ///
+    ///   Note that a future version of this API might panic on pushing
+    ///   `Datum::Null` values.
+    /// - `f` only supports operations which push datums (`push_datum`, etc.).
+    ///   Any other usage results in undefined, unsafe behavior.
+    /// - To construct an empty range, use [`Self::push_empty_range`]. Note
+    ///   that it is possible to generate semantically empty ranges with this
+    ///   function, but they will _not_ be output as empty ranges.
+    ///
+    /// # Panics
+    /// - If any `Datum::Null` were pushed using `f`.
+    /// - If the number of `Datum`s pushed using `f` is incommensurate with the
+    ///   provided `RangeBoundDesc` or the structure of ranges in general.
+    /// - If `Datum`s of different types were pushing using `f`.
+    pub fn push_range_with<F, R>(
+        &mut self,
+        RangeBoundDesc {
+            inf: lower_inf,
+            inc: lower_inc,
+        }: RangeBoundDesc,
+        RangeBoundDesc {
+            inf: upper_inf,
+            inc: upper_inc,
+        }: RangeBoundDesc,
+        f: F,
+    ) -> Result<R, RangeError>
+    where
+        F: FnOnce(&mut RowPacker) -> R,
+    {
+        let start = self.row.data.len();
+        self.row.data.push(Tag::Range.into());
+
+        // write a dummy len, will fix it up later
```

Comment:
Is this supposed to be here now?

Reply:
Ah, this was a good question!

This was supposed to be here because it's using the untagged bytes pattern (i.e. tell us how many bytes belong to this column), but that's actually unnecessary in this case because we're encoding the number of datums we expect in the encoding's flags. I wanted to include the untagged bytes to ensure we weren't mishandling barrier between this column and the next, but realize it's superfluous and blows up the size of the column a good bit.